### PR TITLE
Wenyi5608 greedy sampling

### PR DESCRIPTION
**Pull request by wenyi5608**
Greedy Greedy search added repeat_penalty top_k top_p and temperature function

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
Can you maybe create a generic function which performances sampling and adopts top_p, top_k, num_beams and temperature ? 
Currently, we already have beam search with repetition penalty, but other parameters don't have effect here. 

Ideally, we need a single method similar to HF's generate, which hides generation loop from user. Could you please try it?

**Comment by wenyi5608** [comment]
@ilya-lavrenov HuggingFace's group_beam_search does not implement sampling( top_p, top_k, and temperature) function. Only beam_search and greedy_search have sampling function. genai implements group_beam_search and greedy_search function.

File: text_generation/causal_lm/cpp/greedy_sampling.hpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            // top_p sampling
+            if (0.f < parameters.top_p && parameters.top_p < 1.f) {
+                auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), parameters.top_p);
+                token_scores.resize(pos - token_scores.data());
+            }
+
+            // sample next token
+            sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());
+            for (size_t i = 0; i < token_scores.size(); i++) {
+                logits[i] = token_scores[i].score;
```
Comment (apaniukov) [inline text_generation/causal_lm/cpp/greedy_sampling.hpp:148]: You must take the get the values from logits here, not scores. The softmax already modified the scores in the top_p call, so the second softmax will skew the distribution towards the most probable token.

Comment (apaniukov) [inline text_generation/causal_lm/cpp/greedy_sampling.hpp:148]: You must take the values from logits here, not scores, and then do the softmax on them. The softmax already modified the scores in the top_p call, so the second softmax will skew the distribution towards the most probable token.

Comment (wenyi5608) [reply text_generation/causal_lm/cpp/greedy_sampling.hpp:148]: I refer to chatglm cpp(https://github.com/li-plus/chatglm.cpp/blob/main/chatglm.cpp#L825) to implement this function. 
Is this implementation different from HuggingFace.

Comment (apaniukov) [reply text_generation/causal_lm/cpp/greedy_sampling.hpp:148]: I checked the softmax effects on the logits and the results for the llama model are that the second softmax makes the distribution nearly uniform and the third one makes the distribution exactly uniform. Here is how the probabilities of the most probable token are evolving:
```
Softmax 0 times: tensor([[ 0.0000,  8.6805,  9.4036,  8.4245, 10.8393]])
Softmax 1 times: tensor([[3.1250e-05, 2.7906e-01, 2.0102e-01, 7.2127e-02, 2.6164e-01]])
Softmax 2 times: tensor([[3.1250e-05, 4.1308e-05, 3.8207e-05, 3.3586e-05, 4.0595e-05]])
Softmax 3 times: tensor([[3.1250e-05, 3.1250e-05, 3.1250e-05, 3.1250e-05, 3.1250e-05]])
For uniform distribution: 3.125e-05
```
The second softmax makes model predictions useless. 

Huggingface doesn't transform the logits during `top_p`, just calculates softmax and filters the original logits based on the result. So `sampling_softmax_inplace` is not needed inside the `sampling_top_p` function.


Code for the softmax probs:
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer


model_checkpoint = "JackFram/llama-68m"
hf_tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
hf_model = AutoModelForCausalLM.from_pretrained(model_checkpoint)
texts = ["This is a test"]
tokenized = hf_tokenizer(texts, return_tensors="pt")

with torch.no_grad():
    t = hf_model(**tokenized).logits
    print("Softmax 0 times:", torch.max(t, dim=-1).values)
    for i in range(1, 4):
        t = torch.nn.functional.softmax(t, dim=-1,)
        print(f"Softmax {i} times:", torch.max(t, dim=-1).values)

print(f"For uniform distribution: {1 / hf_tokenizer.vocab_size}")
```

Comment (wenyi5608) [reply text_generation/causal_lm/cpp/greedy_sampling.hpp:148]: @apaniukov Thanks a lot!

Comment (wenyi5608) [reply text_generation/causal_lm/cpp/greedy_sampling.hpp:149]: I re-implemented the sampling_top_p function, the implementation is similar to the implementation of huggingface.

File: text_generation/causal_lm/cpp/greedy_sampling.hpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            thread_local std::mt19937 gen(rd());
+
+            std::discrete_distribution<> dist(logits, logits + token_scores.size());
+            out_token = token_scores[dist(gen)].id;
+        }
+        else {
+            out_token = std::max_element(logits, logits + vocab_size) - logits;
+        }
+
+        prompt.push_back(out_token);
```
Comment (as-suvorov) [inline text_generation/causal_lm/cpp/greedy_sampling.hpp:162]: It seems tokens accumulation needed for applying `sampling_repetition_penalty`. This ids needs to preserved between `get_out_token` calls.

Comment (wenyi5608) [reply text_generation/causal_lm/cpp/greedy_sampling.hpp:162]: Do you mean that the logits info remains the same before and after get_out_token calls?

**Comment by as-suvorov** [comment]
@wenyi5608 , random sampling decoding based on your PR was merged to generate api: https://github.com/pavel-esir/openvino.genai/pull/6. Generate API is preparing for the merge to main repository: https://github.com/openvinotoolkit/openvino.genai/pull/334. Thank you for your contribution!
